### PR TITLE
feat(backend/frontend/shared): user profile screen bb-187

### PR DESCRIPTION
**Pull request by marharita08**
## Summary
Added user profile screen when displayed user data. Added form for updating user data.
### Frontend
- /profile
### Backend
- GET /users/:id
- PATCH /users/:id
## Added
- user profile screen
- update user form
- users redux store including actions:
  - get by id
  - update
## Updated
- user controller: added new endpoints
- user service: implemented update method
- user repository: implemented update method
## Screenshot
image

### CODE REVIEW

File: apps/backend/src/modules/users/user.controller.ts [hunk 1]
```diff
@@ -44,6 +51,32 @@ class UserController extends BaseController {
...
+						params: UserGetParametersDto;
+					}>,
+				),
+			method: "GET",
+			path: UsersApiPath.GET,
+		});
+
+		this.addRoute({
+			handler: (options) =>
+				this.update(
```
Comment (fshabanov) [inline apps/backend/src/modules/users/user.controller.ts:68]: Add a prehandler - check that the user updates their own profile

File: apps/backend/src/modules/users/user.repository.ts [hunk 2]
```diff
@@ -110,8 +110,13 @@ class UserRepository implements Repository {
 			: null;
 	}
 
-	public update(): ReturnType<Repository["update"]> {
-		return Promise.resolve(null);
+	public async update(
+		id: number,
+		payload: Partial<UserDetailsModel>,
+	): Promise<null | UserEntity> {
+		await this.userDetailsModel.query().patch(payload).where({ user_id: id });
```
Comment (fshabanov) [inline apps/backend/src/modules/users/user.repository.ts:117]:
```suggestion
		await this.userDetailsModel.query().patch(payload).where({ userId: id });
```

File: apps/frontend/src/modules/users/libs/types/types.ts [hunk 3]
```diff
@@ -1,9 +1,19 @@
+import { type UserUpdateRequestDto } from "shared";
+
+type UserUpdatePayload = {
```
Comment (fshabanov) [inline apps/frontend/src/modules/users/libs/types/types.ts:3]: Create this type in a separate file

File: apps/frontend/src/pages/profile/libs/components/update-user-form/update-user-form.tsx [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+	user?: UserDto;
+};
+
+const UpdateUserForm: React.FC<Properties> = ({
+	onCancel,
+	onSubmit,
+	user,
+}: Properties) => {
+	const { control, errors, handleSubmit } = useAppForm<UserUpdateRequestDto>({
+		defaultValues: { name: user?.name ?? "" },
```
Comment (fshabanov) [inline apps/frontend/src/pages/profile/libs/components/update-user-form/update-user-form.tsx:23]:
```suggestion
		defaultValues: { name: (user as UserDto).name},
```

File: apps/frontend/src/pages/profile/profile.tsx [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+	const toggleIsUpdate = useCallback((): void => {
+		setIsUpdating(!isUpdating);
+	}, [setIsUpdating, isUpdating]);
+
+	const handleUpdateSubmit = useCallback(
+		(payload: UserUpdateRequestDto): void => {
+			if (id) {
+				void dispatch(usersActions.update({ data: payload, id: +id }));
+			}
```
Comment (fshabanov) [inline apps/frontend/src/pages/profile/profile.tsx:49]: Looks like anyone can update any profile just by changing id in the url. Let's make the route `/profile` and we will display only the current user data

File: apps/frontend/src/pages/profile/profile.tsx [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+		},
+		[dispatch, id],
+	);
+
+	const isLoading = dataStatus === DataStatus.PENDING;
+
+	return (
+		<>
+			{isLoading && <Loader />}
```
Comment (fshabanov) [inline apps/frontend/src/pages/profile/profile.tsx:58]: This is already handled in ProtectedRoute

File: apps/frontend/src/pages/profile/profile.tsx [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+import styles from "./styles.module.css";
+
+const Profile: React.FC = () => {
+	const { id } = useParams();
+	const dispatch = useAppDispatch();
+	const { dataStatus, user } = useAppSelector(({ users }) => ({
+		dataStatus: users.dataStatus,
+		user: users.user,
+	}));
+	const [isUpdating, setIsUpdating] = useState<boolean>(false);
```
Comment (fshabanov) [inline apps/frontend/src/pages/profile/profile.tsx:27]:
```suggestion
```
We don't need that, user can update without clicking on a specific button

**Review by fshabanov — CHANGES_REQUESTED** [review]

File: apps/frontend/src/pages/profile/profile.tsx (hunk 6 above)
Comment (marharita08) [reply apps/frontend/src/pages/profile/profile.tsx:58]: 1. There is no loader in ProtectedRoute. I fixed issue with page loading in app.tsx.
2. Loader in app.tsx is shown when some action in **auth** store is pending, this loader is shown when some action in **users** store is pending.

File: apps/backend/src/libs/types/service.type.ts [hunk 8]
```diff
@@ -5,7 +5,7 @@ type Service<T = unknown> = {
 	findAll(): Promise<{
 		items: T[];
 	}>;
-	update(id: number, payload: unknown): Promise<T>;
+	update(id: number, payload: Partial<T>): Promise<T>;
```
Comment (ihorLysak) [inline apps/backend/src/libs/types/service.type.ts:8]: @fshabanov Should we use Partial<> or unknown here?

Comment (fshabanov) [reply apps/backend/src/libs/types/service.type.ts:8]: I think Partial is better

File: apps/frontend/src/pages/profile/profile.tsx (hunk 6 above)
Comment (fshabanov) [reply apps/frontend/src/pages/profile/profile.tsx:43]: You're right, let's keep it

File: packages/shared/src/modules/users/libs/validation-schemas/user-update.validation-schema.ts [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+
+import { UserValidationMessage, UserValidationRule } from "../enums/enums.js";
+
+type UserUpdateRequestValidationDto = {
+	name: z.ZodString;
+};
+
+const userUpdate = z
+	.object<UserUpdateRequestValidationDto>({
+		name: z.string().trim().min(UserValidationRule.NAME_MINIMUM_LENGTH, {
```
Comment (ihorLysak) [inline packages/shared/src/modules/users/libs/validation-schemas/user-update.validation-schema.ts:11]: Should we check user name for some special characters?

File: apps/frontend/src/pages/profile/libs/components/update-user-form/update-user-form.tsx [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+				control={control}
+				errors={errors}
+				isDisabled
+				label="Email"
+				name="email"
+				placeholder="email"
+				type="email"
+			/>
+			<Button label="SAVE" type="submit" />
+		</form>
```
Comment (ihorLysak) [inline apps/frontend/src/pages/profile/libs/components/update-user-form/update-user-form.tsx:54]: button type is inferred automatically if button is inside of form, more details [here](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#type)

File: packages/shared/src/modules/users/libs/validation-schemas/user-update.validation-schema.ts (hunk 9 above)
Comment (marharita08) [reply packages/shared/src/modules/users/libs/validation-schemas/user-update.validation-schema.ts:11]: I will check requirements for user field.

File: apps/frontend/src/pages/profile/libs/components/update-user-form/update-user-form.tsx (hunk 10 above)
Comment (marharita08) [reply apps/frontend/src/pages/profile/libs/components/update-user-form/update-user-form.tsx:54]: But in Button component default type is 'button' specified.

File: apps/backend/src/libs/modules/controller/base-controller.module.ts [hunk 11]
```diff
@@ -29,9 +30,30 @@ class BaseController implements Controller {
 		this.logger.info(`${request.method.toUpperCase()} on ${request.url}`);
 
 		const handlerOptions = this.mapRequest(request);
-		const { payload, status } = await handler(handlerOptions);
+		const handlerResponse = await handler(handlerOptions);
```
Comment (fshabanov) [inline apps/backend/src/libs/modules/controller/base-controller.module.ts:33]: preHandler should be added to a specific addRoute, no need for mapPreHandler and other logic in this file

File: apps/backend/src/libs/modules/controller/libs/types/api-handler.type.ts [hunk 12]
```diff
@@ -3,6 +3,6 @@ import { type APIHandlerResponse } from "./api-handler-response.type.js";
 
 type APIHandler = (
 	options: APIHandlerOptions,
-) => APIHandlerResponse | Promise<APIHandlerResponse>;
+) => APIHandlerResponse | Promise<APIHandlerResponse> | undefined;
```
Comment (fshabanov) [inline apps/backend/src/libs/modules/controller/libs/types/api-handler.type.ts:6]:
```suggestion
) => APIHandlerResponse | Promise<APIHandlerResponse>;
```

File: apps/backend/src/libs/modules/controller/libs/types/api-pre-handler-response.type.ts [hunk 13]
```diff
@@ -0,0 +1,7 @@
+type APIPreHandlerResponsePayload = {
+	error: Error;
+};
+
+type APIPreHandlerResponse = APIPreHandlerResponsePayload | undefined;
```
Comment (fshabanov) [inline apps/backend/src/libs/modules/controller/libs/types/api-pre-handler-response.type.ts:5]: preHandlers should return void

File: apps/backend/src/libs/modules/controller/libs/types/controller-route-parameters.type.ts [hunk 14]
```diff
@@ -2,11 +2,13 @@ import { type HTTPMethod } from "~/libs/modules/http/http.js";
 import { type ValidationSchema } from "~/libs/types/types.js";
 
 import { type APIHandler } from "./api-handler.type.js";
+import { type APIPreHandler } from "./api-pre-handler.type.js";
 
 type ControllerRouteParameters = {
 	handler: APIHandler;
 	method: HTTPMethod;
 	path: string;
+	preHandler?: APIPreHandler;
```
Comment (fshabanov) [inline apps/backend/src/libs/modules/controller/libs/types/controller-route-parameters.type.ts:11]:
```suggestion
	preHandlers?: APIPreHandler[];
```
We can have multiple

File: apps/backend/src/libs/modules/server-application/base-server-application.ts [hunk 15]
```diff
@@ -132,7 +132,7 @@ class BaseServerApplication implements ServerApplication {
 	}
 
 	public addRoute(parameters: ServerApplicationRouteParameters): void {
-		const { handler, method, path, validation } = parameters;
+		const { handler, method, path, preHandler, validation } = parameters;
```
Comment (fshabanov) [inline apps/backend/src/libs/modules/server-application/base-server-application.ts:135]:
```suggestion
		const { handler, method, path, preHandlers = [], validation } = parameters;
```

File: apps/backend/src/libs/modules/server-application/libs/types/server-application-route-parameters.type.ts [hunk 16]
```diff
@@ -10,6 +10,10 @@ type ServerApplicationRouteParameters = {
 	) => Promise<void> | void;
 	method: HTTPMethod;
 	path: string;
+	preHandler?: (
+		request: FastifyRequest,
+		reply: FastifyReply,
+	) => Promise<void> | void;
```
Comment (fshabanov) [inline apps/backend/src/libs/modules/server-application/libs/types/server-application-route-parameters.type.ts:16]:
```suggestion
		reply: FastifyReply,
		done: HookHandlerDoneFunction
	) => Promise<void> | void;
```

File: apps/backend/src/libs/modules/controller/libs/types/controller-route-parameters.type.ts (hunk 14 above)
Comment (fshabanov) [reply apps/backend/src/libs/modules/controller/libs/types/controller-route-parameters.type.ts:11]: The type should be the same as in server-application-route-parameters.type.ts

File: apps/backend/src/modules/users/user.controller.ts [hunk 17]
```diff
@@ -44,6 +54,52 @@ class UserController extends BaseController {
...
+	private checkUserAccessToUpdate(
+		options: APIHandlerOptions<{
+			params: UserUpdateParametersDto;
+			user: UserDto;
+		}>,
+	): APIPreHandlerResponse {
+		const { params, user } = options;
+
+		if (user.id !== params.id) {
+			return { error: new ForbiddenError({ message: ErrorMessage.FORBIDDEN }) };
```
Comment (fshabanov) [inline apps/backend/src/modules/users/user.controller.ts:101]:
```suggestion
			return { error: new UserError({ message: ErrorMessage.FORBIDDEN }) };
```

Comment (fshabanov) [inline apps/backend/src/modules/users/user.controller.ts:101]: Just throw an error

File: apps/frontend/src/modules/users/slices/actions.ts [hunk 18]
```diff
@@ -0,0 +1,25 @@
+import { createAsyncThunk } from "@reduxjs/toolkit";
+
+import { type AsyncThunkConfig } from "~/libs/types/types.js";
+import { type UserDto, type UserUpdatePayload } from "~/modules/users/users.js";
+
+import { name as sliceName } from "./users.slice.js";
+
+const getUserFromAuth = createAsyncThunk<UserDto, undefined, AsyncThunkConfig>(
```
Comment (fshabanov) [inline apps/frontend/src/modules/users/slices/actions.ts:8]: Why do we need this as an action?

File: apps/frontend/src/pages/profile/libs/components/update-user-form/update-user-form.tsx [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+	type UserUpdateFormDto,
+	type UserUpdateRequestDto,
+	userUpdateValidationSchema,
+} from "~/modules/users/users.js";
+
+import styles from "./styles.module.css";
+
+type Properties = {
+	onSubmit: (payload: UserUpdateRequestDto) => void;
+	user?: UserDto;
```
Comment (fshabanov) [inline apps/frontend/src/pages/profile/libs/components/update-user-form/update-user-form.tsx:14]:
```suggestion
	user: UserDto;
```

**Review by fshabanov — CHANGES_REQUESTED** [review]

File: apps/frontend/src/modules/users/slices/actions.ts (hunk 18 above)
Comment (marharita08) [reply apps/frontend/src/modules/users/slices/actions.ts:8]: I have to get user from auth store and put it into users store. What a correct way to do it?

Comment (fshabanov) [reply apps/frontend/src/modules/users/slices/actions.ts:8]: Use a user from auth slice directly in the component, no need to move it to user slice

Comment (marharita08) [reply apps/frontend/src/modules/users/slices/actions.ts:8]: then how should I update user in users slice, not moving user to this slice

Comment (fshabanov) [reply apps/frontend/src/modules/users/slices/actions.ts:9]: Let's create a new endpoint getById and get the user from there. Don't forget to add a preHandler check

File: apps/backend/src/libs/modules/controller/base-controller.module.ts [hunk 20]
```diff
@@ -34,6 +35,16 @@ class BaseController implements Controller {
 		return await reply.status(status).send(payload);
 	}
 
+	private mapPreHandler({
```
Comment (fshabanov) [inline apps/backend/src/libs/modules/controller/base-controller.module.ts:38]: We don't need mapPreHandler. We should add preHandlers directly to routeParameters in `base-server-application.ts` addRoute method. The preHandler function will receive request, reply, done and proceed with the logic

**Review by fshabanov — CHANGES_REQUESTED** [review]

File: apps/backend/src/libs/modules/controller/base-controller.module.ts (hunk 20 above)
Comment (marharita08) [reply apps/backend/src/libs/modules/controller/base-controller.module.ts:38]: 1. In user controller I created pre handler function with type `(options: APIHandlerOptions) => void`.
2. In base controller I use mapPreHandler to convert preHandlers from `(options: APIHandlerOptions) => void` to `(request: FastifyRequest, reply: FastifyReply, done: HookHandlerDoneFunction,) => void`.
3. In base server application in addRoute I transfered all preHandlers to app.route.
Tried to do everything as with handles.
I don't understand, should I create function with `(request: FastifyRequest, reply: FastifyReply, done: HookHandlerDoneFunction,) => void` type right in user controller?

Comment (fshabanov) [reply apps/backend/src/libs/modules/controller/base-controller.module.ts:38]: Yes, you should create function with `(request: FastifyRequest, reply: FastifyReply, done: HookHandlerDoneFunction,) => void` type. You should create it in a separate file though

File: apps/backend/src/libs/modules/server-application/base-server-application.ts [hunk 21]
```diff
@@ -132,11 +132,12 @@ class BaseServerApplication implements ServerApplication {
 	}
 
 	public addRoute(parameters: ServerApplicationRouteParameters): void {
-		const { handler, method, path, validation } = parameters;
+		const { handler, method, path, preHandlers = [], validation } = parameters;
 
 		this.app.route({
 			handler,
 			method,
+			preHandler: [...preHandlers],
```
Comment (fshabanov) [inline apps/backend/src/libs/modules/server-application/base-server-application.ts:140]:
```suggestion
			preHandler: preHandlers,
```

**Review by fshabanov — CHANGES_REQUESTED** [review]

File: apps/backend/src/modules/users/libs/hooks/check-access-to-user-data.pre-handler.ts [hunk 22]
```diff
@@ -0,0 +1,23 @@
+import { ErrorMessage } from "~/libs/enums/enums.js";
```
Comment (fshabanov) [inline apps/backend/src/modules/users/libs/hooks/check-access-to-user-data.pre-handler.ts:1]: Change file name to `check-access-to-user-data.hook.ts`

**Review by fshabanov — CHANGES_REQUESTED** [review]

**Review by fshabanov — APPROVED** [review]

**Pull request merged**